include/mlir/Analysis/Liveness.h

```diff
@@ -0,0 +1,131 @@
+// =============================================================================
+//
+// This file contains an analysis for computing liveness information from a
+// given top-level operation. The current version of the analysis uses a
+// traditional algorithm to resolve detailed live-range information about all
+// values within the specified regions. It is also possible to query liveness
+// information on block level.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_ANALYSIS_LIVENESS_H
+#define MLIR_ANALYSIS_LIVENESS_H
+
+#include <vector>
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/BitVector.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/SetVector.h"
+
+namespace mlir {
+class Block;
+class Region;
+class Value;
+class Operation;
+
+class LivenessBlockInfo;
+
+/// Represents an analysis for computing liveness information from a
+/// given top-level operation.
+class Liveness {
+public:
+  using OperationIdMapT = llvm::DenseMap<Operation *, size_t>;
+  using OperationListT = std::vector<Operation *>;
+  using ValueMapT = llvm::DenseMap<Value *, llvm::BitVector>;
+  using BlockMapT = llvm::DenseMap<Block *, LivenessBlockInfo>;
+  using ValueSetT = llvm::SmallSetVector<Value *, 16>;
+
+public:
+  /// Creates a new Liveness analysis that computes liveness
+  /// information for all associated regions.
+  Liveness(Operation *op);
+
+  /// Returns the operation this analysis was constructed from.
+  Operation *getOperation() const { return operation; }
+
+  /// Resolves liveness info (if any) for the given value.
+  /// This includes all operations in which the given value is live.
+  OperationListT resolveLiveness(Value *value) const;
+
+  /// Resolves liveness info (if any) for the block.
+  const LivenessBlockInfo *getLiveness(Block *block) const;
+
+  /// Returns a reference to a set containing live-in values.
+  const ValueSetT &getLiveIn(Block *block) const;
+
+  /// Returns a reference to a set containing live-out values.
+  const ValueSetT &getLiveOut(Block *block) const;
+
+  /// Returns true if the given operation (and its associated operand index)
+  /// represent the last use of the given value.
+  bool isLastUse(Value *value, Operation *operation, unsigned operandIndex);
+
+  /// Dumps the liveness information in a human readable format.
+  void dump() const;
+
+  /// Dumps the liveness information to the given stream.
+  void print(llvm::raw_ostream &os) const;
+
+private:
+  /// Initializes the internal mappings.
+  void build(llvm::MutableArrayRef<Region> regions);
+
+private:
+  Operation *operation;
+
+  /// Maps operations to unique identifiers (for fast bit-vector lookups).
+  OperationIdMapT operationIdMapping;
+
+  /// Ordered list of all operations that can be used to map bit indices
+  /// (from bit-vector lookups) to actual operations.
+  OperationListT operationList;
+
+  /// Maps values to internal liveness information.
+  ValueMapT valueMapping;
```

Comment:
> This requires a "backward" mapping of operation IDs to operations. 

Right, but could we keep storage of the liveness of "blocks" only for instance and populate the set on-demand by traversing the blocks. Basically space could be N^2 in terms of number of blocks instead of number of instructions.

I haven't really thought about it, so I may be totally off here, but some rough idea of what crossed my mind was to keep only the "reachability" between blocks and then do something along the line of the following for `isLastUse` (super approximated code):

```
bool Liveness::isLastUse(Value *value, Operation *operation) {
  // operation is not the last use if we find another user live after operation.
  for (Operation *user : value->getUsers()) {
    if (isLiveAfter(user, operation)) // returns true if `user` is live after `operation`
      return false;
  return true;
}

bool Liveness::isLiveAfter(Operation *op1, Operation *op2) {
   // in the same block, op2 is live after op1 if sequentially after.
   if (op1->getBlock() == op2->getBlock() && op2->isAfterInBlock(op1)) return true;
   // If there is a patch from the block holding op1 to the blocking holding op2, then op2 is live after op1
   return isReachable(op1->getBlock(), op2->getBlock()); // just encode that there is a patch between the blocks
}
```

Reply:
I think this is out of scope of the current PR. I would propose to discuss that in `mlir@tensorflow.org` and then submit a follow-up PR that adds this behaviour.

Reply:
I don’t know what you mean here, can you clarify what is out of scope and why?

Reply:
Because this sounds like a good optimization that can be submitted in a follow-up PR.

Reply:
I disagree: this seems absolutely the kind of thing go discuss and address during code review.

Reply:
I agree that we could compute the set of live operations as needed and thus on-the-fly upon request. What do you mean by N^2 in terms of the number of blocks? Do you suggest that we should store reachability information to allow a constant time evaluation of `isReachable`?
In addition, we could expose the `getStartOperation` and `getEndOperation` functions to determine the start or end operation of a particular `value` within a `block`. This requires neither bit sets nor complex lookup tables. This would allow fine-grained liveness intervals to be determined in addition to the proposed `isLastUse` functionality.